R/preprocessing.R

```diff
@@ -1526,3 +1526,165 @@ RegressOutNB <- function(
   object@scale.data <- pr
   return(object)
 }
+
+
+
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+# Proposed feature: subset by barcode distribution inflection points
+#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+#' Subsets a Seurat Object By Distribution of UMIs
+#'
+#' This function calculates an adaptive inflection point ("knee") of the barcode distribution
+#' for each sample group (as identified by `object@meta.data$orig.ident`).
+#' Returns the subsetted `object` with cells that passed the calculated threshold.
+#'
+#' The assumed assay is "RNA", and thus uses the "nCount_RNA" distribution for
+#' calculating the appropriate cutoff and returning a subsetted object.
+#'
+#' @param object Seurat object
+#' @param umi.column Column to use as proxy for barcodes ("nCount_RNA" by default)
+#' @param show.plot Whether to produce diagnostic plot showing inflection point and barcode distribution
+#' @param threshold.low Ignore barcodes of rank below this threshold in inflection calculation
+#' @param threshold.high Ignore barcodes of rank above this threshold in inflection calculation
+#' @param return.plot Returns the ggplot2 object in lieu of the subsetted Seurat object (useful for visually testing thresholds)
+#' @param return.cells Returns the vector of cells which pass the defined barcode thresholds
+#' @param return.inflections Returns a tibble with the grouping variable, data list-column with cell, umi.column, and rank, the calculated inflection point and associated umi_threshold, and a list-column of the cells which passed the calculated threshold per group
+#'
+#' @return Return Seurat object subsetted. Appends to the workflow slot 'nCount_RNA_threshold'
+#' with a `data.frame` showing the calculated thresholds per sample group that determine the passing cells.
+#'
+#' @importFrom dplyr mutate arrange group_by summarise select inner_join filter
+#' @importFrom tidyr nest
+#' @importFrom purrr map_int map2_dbl map2
+#' @importFrom ggplot2 ggplot geom_line theme_bw geom_vline
+#' @importFrom rlang sym
+#' @import tibble
```

Comment:
One thing we're working on with Seurat v3 is reducing general `@import` in favor of explicit `@importFrom`; could you specify what from tibble you're importing here?

Reply:
Great, will do. One question re: the use of `tibble` though. The reason I use it is to organize some data into "list-columns" for easier access/annotation of data, which requires the use of tibble to view neatly (aka, without it printing a bunch of lines in an unreadable way). It seems like Seurat v3 doesn't really use the `tibble` format in general, so should I just get rid of this dependency and use `data.frame` instead? I personally prefer `tibble` to `data.frame` (better printing, list-columns are great), but will go with what you think is best.